SignalFx exporter: Add host metadata synchronization

dmitryax opened the pull request:
**Description:** 
This change introduces host metadata updates functionality in Signalfx exporter. Synchronization happens once first metric batch is received. It looks for dimensions set by resourcedetection processor

dmitryax commented:
as discussed with @jrcamp it will be changes to use cloud unique id instead of host

jrcamp commented:
@dmitryax I think the PR has become too big. I think it'd be good to break it up at least into two pieces, one being stuff that came from smart agent that's already been vetted there and the second is new code that may need a closer look.

dmitryax commented:
@jrcamp broke it down as you suggested. The first part: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1118

dmitryax commented on exporter/signalfxexporter/hostmetadata/metadata.go:
```diff
@@ -0,0 +1,145 @@
+	}
+
+	metadataUpdate := s.prepareMetadataUpdate(props, hostID)
+	err := s.dimClient.PushMetadata([]*collection.MetadataUpdate{metadataUpdate})
+	if err != nil {
+		s.logger.Error("Failed to push host metadata update", zap.Error(err))
+		return
+	}
```
thanks. i'll do that using sync.Once

dmitryax commented on exporter/signalfxexporter/hostmetadata/metadata.go:
```diff
@@ -0,0 +1,145 @@
+	dimClient dimensions.MetadataUpdateClient
+	synced    bool
+
+	done chan struct{}
```
not anymore. good catch, thanks.

dmitryax replied:
fixed

dmitryax commented on exporter/signalfxexporter/hostmetadata/metadata.go:
```diff
@@ -0,0 +1,145 @@
+
+var (
+	// Resource attribute that used as fallback if there are no cloud provider attributes.
+	extraHostIDAttribute = "host"
```
that makes sense. I'll update it in this PR

dmitryax replied:
updated

jrcamp commented on exporter/signalfxexporter/go.mod:
```diff
@@ -15,6 +15,7 @@ require (
 	go.uber.org/multierr v1.5.0
 	go.uber.org/zap v1.16.0
 	google.golang.org/protobuf v1.25.0
+	gotest.tools v2.2.0+incompatible
```
Use testify assert instead of adding new dep.

dmitryax replied:
oh nice catch! I got it automatically imported

jrcamp replied:
Yeah it's always importing the wrong one for me haha. Major pain.

jrcamp commented:
lgtm after the assert thing is fixed and #1100 is merged (should be soon) so that can be removed here.

dmitryax committed:
SignalFx exporter: Add periodic host metadata updates syncronization

This change introduces host metadata updates functionality in Signalfx exporter. Synchronization happens once first metric batch is received. It looks for dimensions set by resourcedetection processor

dmitryax commented:
@jrcamp , PR rebased
